File: C-Plus-Plus/BinarySearch.cpp
#include <bits/stdc++.h>
using namespace std;

int Binarysearch(int arr[], int target, int l, int r)
{
    while (l <= r)
    {
        int m = l + (r - l) / 2;
        if (arr[m] == target)
        {
            return m;
        }

        if (arr[m] < target)
        {
            l = m + 1;
        }
        else
        {
            r = m - 1;
        }
    }
    return -1;
}

int main()
{
    int n;
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    int target;
    cin >> target;
    int index_found = Binarysearch(arr, target, 0, n - 1);
    if (index_found == -1)
    {
        cout << "The number is not found\n";
    }
    else
    {
        cout << "The number is found at index " << index_found << endl;
    }
    return 0;
}

/*
Time Complexity : O(logn)
Space Complexity : O(n)

User is required to enter the size of array and then elements of array
in a space seperated manner. At last the user has to enter the target value 
to be searched.

Input : 
5
1 2 3 4 5
3

Output : The number is found at index 2

Input :
5
1 2 3 4 5
6

Output : The number is not found
*/

File: JavaScript/binarySearch.js
const binarySearch = (sortedArray, seekElement) => {
  let startIndex = 0;
  let endIndex = sortedArray.length - 1;

  while (startIndex <= endIndex) {
    const middleIndex = startIndex + Math.floor((endIndex - startIndex) / 2);
    if (sortedArray[middleIndex] === seekElement) {
      return middleIndex;
    }
    if (sortedArray[middleIndex] < seekElement) {
      startIndex = middleIndex + 1;
    } else {
      endIndex = middleIndex - 1;
    }
  }

  return -1;
};

const sortedArray = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ];
const firstSeekElement = 8;
const secondSeekElement = 21;
console.log(binarySearch(sortedArray, firstSeekElement)); // Returns 7, index of seekElement;
console.log(binarySearch(sortedArray, secondSeekElement)); // Returns -1 because secondSeekElement was not found in the given array.

/*
 * Input:
 *    sortedArray: Array - Sorted array, e.g.: [ 1, 2, 3, 4, 5 ]
 *    seekElement: Any - Element that we are searching for in sortedArray.
 * Output:
 *    Number - The index of the searched element or -1 if the element was not found.
 *
 * Time Complexity: O(log(n))
 * Space Complexity: O(1)
*/
